### PR TITLE
Fixing AFA Pipe

### PR DESCRIPTION
**Pull request by rodrigoaatmicrosoft**
This change contains 3 changes to AFA pipe single event
1- SyncTimeSimultaneityFree
2- Stale state
3- Event partition

### CODE REVIEW

**Review by peterfreiling — DISMISSED** [review]
:shipit:

File: Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs [hunk 1]
```diff
@@ -274,21 +273,19 @@ public override unsafe void OnNext(StreamMessage<TKey, TPayload> batch)
...
                                                 {
-                                                    var otherTime = Math.Min(synctime + this.MaxDuration, StreamEvent.InfinitySyncTime);
-
                                                     if (!this.IsSyncTimeSimultaneityFree)
                                                     {
                                                         var tentativeOutputEntry = this.tentativeOutput.entries[partitionIndex].value;
 
                                                         int ind = tentativeOutputEntry.Insert(src_hash[i]);
-                                                        tentativeOutputEntry.Values[ind].other = otherTime;
+                                                        tentativeOutputEntry.Values[ind].other = synctime + this.MaxDuration;
```
Comment (arunkm) [inline Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs:281]: Can `synctime + MaxDuration` overshoot MaxDate ?

File: Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs [hunk 2]
```diff
@@ -72,16 +70,16 @@ public override unsafe void OnNext(StreamMessage<TKey, TPayload> batch)
...
 
                             if (!this.IsSyncTimeSimultaneityFree)
                             {
                                 int index;
+                                var tentativeVisibleTraverser = new FastMap<OutputEvent<TKey, TRegister>>.VisibleTraverser(this.tentativeOutput.entries[partitionIndex].value);
 
                                 if (synctime > this.lastSyncTime.entries[partitionIndex].value) // move time forward
                                 {
-                                    foreach (var mapIndex in this.seenPartitions.entries[partitionIndex].value) this.seenEvent.Remove(mapIndex);
+                                    this.seenEvent.Remove(srckey[i]);
```
Comment (arunkm) [inline Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs:82]: Considering srckey[i] is assigned to variable key, it is better to use consistently variable `key` throughout

File: Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs (hunk 1 above)
Comment (arunkm) [reply Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs:281]: This undoes the fix done in commit 5d2dfe78

**Review by peterfreiling — DISMISSED** [review]
:clock1:

File: Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs (hunk 1 above)
Comment (rodrigoaatmicrosoft) [reply Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs:281]: Not really, MaxDuration is up to 7 days, synctime is the time now on timestamp. If we are close to year 9998 yes but that is not the case :)

File: Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs (hunk 2 above)
Comment (rodrigoaatmicrosoft) [reply Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs:82]: Changed

File: Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs (hunk 1 above)
Comment (rodrigoaatmicrosoft) [reply Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs:281]: That is a good question, should we continue using the fix on 5d2dfe7 (https://github.com/microsoft/Trill/commit/5d2dfe7815ac883d3ddeea79091a6df50ec6443e). @peterfreiling

Comment (peterfreiling) [reply Sources/Core/Microsoft.StreamProcessing/Operators/Afa/CompiledPartitionedAfaPipe_SingleEvent.cs:281]: Why would we want to revert 5d2dfe7? Please keep 5d2dfe7 .

---
In reply to: [493041257](https://github.com/microsoft/Trill/pull/134#discussion_r493041257) [](ancestors = 493041257)

**Comment by rodrigoaatmicrosoft** [comment]
> Would like to know if the MaxDuration fix change is undone unintentionally or is not required anymore

I added the removed part back again.

**Review by arunkm — APPROVED** [review]

**Review by peterfreiling — APPROVED** [review]

**Pull request merged**